pkg/gitprovider/azure-devops.go

```diff
@@ -0,0 +1,483 @@
+
+	if strings.HasPrefix(repoUrl, "http") {
+		return g.parseAzureDevopsHttpGitUrl(repoUrl)
+	}
+
+	return nil, errors.New("can not parse git URL: " + repoUrl)
+}
+
+func (g *AzureDevOpsGitProvider) parseAzureDevopsSshGitUrl(gitURL string) (*StaticGitContext, error) {
+	re := regexp.MustCompile(`git@ssh.([\w\.]+):(.+?)/(.+?)/(.+?)/(.+?)?$`)
+	matches := re.FindStringSubmatch(gitURL)
+	if len(matches) != 6 {
+		return nil, errors.New("cannot parse git URL: " + gitURL)
+	}
+
+	repo := &StaticGitContext{}
+
+	repo.Source = matches[1]
+	repo.Owner = matches[3]
+	repo.Name = matches[5]
+	project := matches[4]
+	repo.Url = g.getAzureDevopsCloneUrl(repo.Source, repo.Owner, repo.Name, project)
+	repoId, err := g.getAzureDevopsRepoId(repo.Name, project)
+	if err != nil {
+		return nil, errors.New("cannot parse git URL: " + gitURL)
+	}
+	repo.Id = repoId
+
+	return repo, nil
+}
+
+func (g *AzureDevOpsGitProvider) parseAzureDevopsHttpGitUrl(gitURL string) (*StaticGitContext, error) {
+	u, err := url.Parse(gitURL)
+	if err != nil {
+		return nil, err
+	}
+
+	repo := &StaticGitContext{}
+	urlPattern := `^(https?://)?(?P<source>[^/]+)/(?P<org>[^/]+)(?:/(?P<project>[^/_]+))?/_git/(?P<repo>[^/?]+)(?:\?.*)?(?:/.*)?$`
+	urlPatternRegex := regexp.MustCompile(urlPattern)
+	matches := urlPatternRegex.FindStringSubmatch(gitURL)
+	if len(matches) < 6 {
+		return nil, errors.New("cannot parse git URL: " + gitURL)
+	}
+
+	repo.Source = u.Host
+	repo.Owner = matches[3]
+	repo.Name = matches[5]
+	projectName := matches[4]
+
+	urlPath := strings.TrimPrefix(u.Path, "/")
+
+	if projectName == "" {
+		projectName = repo.Name
+		parts := strings.SplitN(urlPath, fmt.Sprintf("%s/", repo.Owner), 2)
+		if len(parts) != 2 {
+			return nil, errors.New("cannot parse git URL: " + gitURL)
+		}
+
+		urlPath = strings.Join([]string{parts[0], repo.Owner, projectName, parts[1]}, "/")
+		urlPath = strings.TrimPrefix(urlPath, "/")
+	}
+
+	parts := strings.Split(urlPath, "/")
+
+	repo.Url = g.getAzureDevopsCloneUrl(repo.Source, repo.Owner, repo.Name, projectName)
+	repo.Name, _ = url.QueryUnescape(repo.Name)
+	projectName, _ = url.QueryUnescape(projectName)
+
+	repoId, err := g.getAzureDevopsRepoId(repo.Name, projectName)
+	if err != nil {
+		return nil, errors.New("cannot parse git URL: " + gitURL)
+	}
+	repo.Id = repoId
+
+	queryParams, err := url.QueryUnescape(u.RawQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(parts) <= 4 && queryParams == "" {
+		return repo, nil
+	}
+
+	switch {
+	case len(parts) >= 6 && parts[4] == "pullrequest":
+		prNumber, _ := strconv.Atoi(parts[5])
+		prUint := uint32(prNumber)
+		repo.PrNumber = &prUint
+		repo.Path = nil
+	case len(parts) >= 6 && parts[4] == "commit":
+		repo.Sha = &parts[5]
+		repo.Branch = &parts[5]
+		repo.Path = nil
+	}
+
+	switch {
+	case strings.Contains(queryParams, "itemVersion=GB"):
+		fallthrough
+	case strings.Contains(queryParams, "version=GB"):
+		fallthrough
+	case strings.Contains(queryParams, "refName=refs/heads/"):
```

Comment:
@tarunrajput there seems to be an issue with creating a project using a commit URL (e.g. https://dev.azure.com/OWNER/nospacetry/_git/nospacetry/commit/COMMIT?refName=refs%2Fheads%2Fmain).

I can see that your test case for this is correct but the "issue" here happens when the query param is present so this second `switch` overrides the branch.

The expected behavior is that after parsing the context of that URL, the `sha` and `branch` are set to the same value.

Reply:
Oh! I thought Sha and branch can be different. I'll fix this in next 1 hr.

Reply:
The reason they need to be the same in this scenario is so we know that the user intended to clone the repo on that specific commit. 

Side note: https://github.com/daytonaio/daytona/issues/701 currently tracks an issue of cloning from a commit not on the default branch so, currently, you can only test this by using a commit URL from the default branch.

Reply:
fixed